src/base/random.test.cc

```diff
@@ -0,0 +1,112 @@
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <gtest/gtest-spi.h>
+#include <gtest/gtest.h>
+
+#include <unordered_map>
+
+#include "base/gtest/logging.hh"
+#include "base/random.hh"
+
+using namespace gem5;
+
+/**
+ * Checking that default construction uses the default
+ * seed as specified by the standard
+ */
+TEST(Random, DefaultConstruct)
+{
+#ifdef NDEBUG
+    GTEST_SKIP() << "Skipping as assertions are "
+        "stripped out of fast builds";
+#endif
+    // Init seed was default seed 5489
+    Random dut;
+    // First number corresponds to the default seed
+    ASSERT_EQ(dut.random<uint64_t>(), 14514284786278117030llu);
+}
+
+
+/**
+ * Checking that default construction uses the default
+ * see as specified by the standard
+ */
+TEST(Random, ConstructUserSpecifiedSeed)
+{
+#ifdef NDEBUG
+    GTEST_SKIP() << "Skipping as assertions are "
+        "stripped out of fast builds";
+#endif
+    Random dut{42};
+    ASSERT_EQ(dut.random<uint64_t>(), 13930160852258120406llu);
+}
+
+/**
+ * Test that reseeding after construction with
+ * an explicit seed works
+ */
+TEST(Random, ConstructThenReseed)
+{
+#ifdef NDEBUG
+    GTEST_SKIP() << "Skipping as assertions are "
+        "stripped out of fast builds";
+#endif
+
+    Random dut{};
+    ASSERT_EQ(dut.random<uint64_t>(), 14514284786278117030llu);
+    dut.init(42);
+    ASSERT_EQ(dut.random<uint64_t>(), 13930160852258120406llu);
+}
+
+/**
+ * This tests check that range generation
+ * does generate all possible values within
+ * a range
+ */
+TEST(Random, RangeCoverage)
+{
+#ifdef NDEBUG
+    GTEST_SKIP() << "Skipping as assertions are "
+        "stripped out of fast builds";
+#endif
+    Random dut;
+
+    // Count occurences if we want to check
+    // frequencies in the future
```

Comment:
Out of curiosity, how well did the values get distributed? Can we add checks for that, or is that too limiting?

Reply:
Looping 500 times it's 31%/38.8%/30.2%, looping 20k times it's 34%/33.3%/32.7% so within 1%. Looping 20k is fast enough, I added the frequency test.

Another tidbit is that range random can give you negative numbers if you want, whereas the scalar random is limited to positive integers.